signingscript/sign.py

```diff
@@ -228,18 +234,95 @@ def build_signtool_cmd(context, from_, fmt, to=None):
             await sign_file(context, from_, fmt)
     if file_extension == '.zip':
         # Recreate the zipfile
-        await _create_zipfile(context, orig_path, files)
+        await _create_zipfile(context, orig_path, files, tmp_dir=tmp_dir)
     return orig_path
 
 
 # sign_widevine {{{1
 async def sign_widevine(context, orig_path, fmt):
-    """Sign the internals of a file with the widevine key.
+    """Call the appropriate helper function to do widevine signing.
 
-    Extract the zip or tarball and only sign a handful of files (see
+    Args:
+        context (SigningContext): the signing context
+        orig_path (str): the source file to sign
+        fmt (str): the format to sign with
+
+    Raises:
+        SigningScriptError: on unknown suffix.
+
+    Returns:
+        str: the path to the signed archive
+
+    """
+    ext_to_fn = {
+        '.zip': sign_widevine_zip,
+        '.tar.bz2': sign_widevine_tar,
+        '.tar.gz': sign_widevine_tar,
+    }
+    for ext, signing_func in ext_to_fn.items():
+        if orig_path.endswith(ext):
+            return await signing_func(context, orig_path, fmt)
+    raise SigningScriptError(
+        "Unknown widevine file format for {}".format(orig_path)
+    )
+
+
+async def sign_widevine_zip(context, orig_path, fmt):
+    """Sign the internals of a zipfile with the widevine key.
+
+    Extract the files to sign (see `_WIDEVINE_BLESSED_FILENAMES` and
+    `_WIDEVINE_UNBLESSED_FILENAMES), skipping already-signed files.
+    The blessed files should be signed with the `widevine_blessed` format.
+    Then append the sigfiles to the zipfile.
+
+    Args:
+        context (SigningContext): the signing context
+        orig_path (str): the source file to sign
+        fmt (str): the format to sign with
+
+    Returns:
+        str: the path to the signed archive
+
+    """
+    # This will get cleaned up when we nuke `work_dir`. Clean up at that point
+    # rather than immediately after `sign_widevine`, to optimize task runtime
+    # speed over disk space.
+    tmp_dir = tempfile.mkdtemp(prefix="wvzip", dir=context.config['work_dir'])
+    # Get file list
+    all_files = await _get_zipfile_files(orig_path)
+    files_to_sign = _should_sign_widevine(all_files)
+    sig_files = []
+    log.debug("Widevine files to sign: {}".format(files_to_sign))
+    if files_to_sign:
```

Comment:
I don't think we should. I'm hoping we get zero files to sign regularly, e.g. when we sign l10n repacks of signed en-US. We will get a `Widevine files to sign: {}` debug message in that case.